org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java

```diff
@@ -582,6 +568,30 @@ private static boolean isComplianceJava9OrHigher(IJavaProject javaProject) {
 	return CompilerOptions.versionToJdkLevel(javaProject.getOption(JavaCore.COMPILER_COMPLIANCE, true)) >= ClassFileConstants.JDK9;
 }
 
+private static boolean isOnModulePath(JavaProject javaProject, PackageFragmentRoot root) {
+	boolean isOnModulePath = true; // if an exception occurs, assume yes
+	try {
+		IClasspathEntry classpathEntry = root.getRawClasspathEntry();
+		if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+			/*
+			 * Source classpath entries of a project never have the module attribute set,
+			 * so we cannot rely on the attribute.
+			 */
+			isOnModulePath = isModularProject(javaProject);
+		/*
+		 * The JRE container is on the module path based on its version,
```

Comment:
Note, this is not correct assumption and comment. There can be different containers on classpath, not necessarily `JRE_CONTAINER`. I first was tempted to say we should check `if (classpathEnty instanceof IModulePathEntry)` but it seems that `IModulePathEntry` can also represent non-modular code.

So far I see no reliable way to detect if some container entry is on module path...

May be we could use given `PackageFragmentRoot`. Obviously, if that is `instanceof JrtPackageFragmentRoot` it will be always modular. 

But I'm not sure what kind of other `PackageFragmentRoot` objects can be contributed by `CPE_CONTAINER` entries that have no "modular" attribute but still host `PackageFragmentRoot` from modular world.

Reply:
@trancexpress : the comment hast to be changed, it is not correct.

Something like "we have no idea what the container provides, so we assume it contributes something to module path"

Reply:
Also please move the comment to the **corresponding** if case,